File: ios/VllowDugout/Models/ChatMessage.swift
import Foundation

nonisolated struct ChatMessage: Identifiable, Hashable, Sendable {
    let id: UUID
    let role: MessageRole
    let content: String
    let timestamp: Date
    let drillAttachment: DrillAttachment?
    let suggestedQuestions: [String]

    init(
        id: UUID = UUID(),
        role: MessageRole,
        content: String,
        timestamp: Date = Date(),
        drillAttachment: DrillAttachment? = nil,
        suggestedQuestions: [String] = []
    ) {
        self.id = id
        self.role = role
        self.content = content
        self.timestamp = timestamp
        self.drillAttachment = drillAttachment
        self.suggestedQuestions = suggestedQuestions
    }
}

nonisolated enum MessageRole: String, Sendable {
    case user
    case assistant
}

nonisolated struct DrillAttachment: Hashable, Sendable {
    let title: String
    let subtitle: String
    let imageURL: String
    let duration: String
    let linkedDrillID: UUID?

    init(title: String, subtitle: String, imageURL: String, duration: String, linkedDrillID: UUID? = nil) {
        self.title = title
        self.subtitle = subtitle
        self.imageURL = imageURL
        self.duration = duration
        self.linkedDrillID = linkedDrillID
    }

    static func fromDrill(_ drill: Drill) -> DrillAttachment {
        DrillAttachment(
            title: drill.title,
            subtitle: "\(drill.category.rawValue.uppercased()) • \(drill.durationFormatted)",
            imageURL: drill.imageURL,
            duration: drill.durationFormatted,
            linkedDrillID: drill.id
        )
    }
}

File: flutter_app/lib/models/chat_message.dart
import 'package:uuid/uuid.dart';
import 'drill.dart';

enum MessageRole {
  user,
  assistant,
}

class DrillAttachment {
  final String title;
  final String subtitle;
  final String imageURL;
  final String duration;
  final String? linkedDrillID;

  DrillAttachment({
    required this.title,
    required this.subtitle,
    required this.imageURL,
    required this.duration,
    this.linkedDrillID,
  });

  factory DrillAttachment.fromDrill(Drill drill) {
    return DrillAttachment(
      title: drill.title,
      subtitle:
          '${drill.category.displayName.toUpperCase()} • ${drill.durationFormatted}',
      imageURL: drill.imageURL,
      duration: drill.durationFormatted,
      linkedDrillID: drill.id,
    );
  }
}

class ChatMessage {
  final String id;
  final MessageRole role;
  final String content;
  final DateTime timestamp;
  final DrillAttachment? drillAttachment;
  final List<String> suggestedQuestions;

  ChatMessage({
    String? id,
    required this.role,
    required this.content,
    DateTime? timestamp,
    this.drillAttachment,
    this.suggestedQuestions = const [],
  })  : id = id ?? const Uuid().v4(),
        timestamp = timestamp ?? DateTime.now();

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is ChatMessage && id == other.id;

  @override
  int get hashCode => id.hashCode;
}
